java/ql/lib/semmle/code/java/security/UnsafeCertTrust.qll

```diff
@@ -0,0 +1,102 @@
+/** Provides classes and predicates to reason about unsafe certificate trust vulnerablities. */
+
+import java
+private import semmle.code.java.frameworks.Networking
+private import semmle.code.java.security.Encryption
+private import semmle.code.java.dataflow.DataFlow
+
+/**
+ * The creation of an object that prepares an SSL connection.
+ *
+ * This is a source for `SslEndpointIdentificationFlowConfig`.
+ */
+class SslConnectionInit extends DataFlow::Node {
+  SslConnectionInit() {
+    exists(MethodAccess ma, Method m |
+      this.asExpr() = ma and
+      ma.getMethod() = m
+    |
+      m instanceof CreateSslEngineMethod
+      or
+      m instanceof CreateSocketMethod and isSslSocket(ma)
+    )
+  }
+}
+
+/**
+ * A call to a method that establishes an SSL connection.
+ *
+ * This is a sink for `SslEndpointIdentificationFlowConfig`.
+ */
+class SslConnectionCreation extends DataFlow::Node {
+  SslConnectionCreation() {
+    exists(MethodAccess ma, Method m |
+      m instanceof BeginHandshakeMethod or
+      m instanceof SslWrapMethod or
+      m instanceof SslUnwrapMethod or
+      m instanceof SocketGetOutputStreamMethod
+    |
+      ma.getMethod() = m and
+      this.asExpr() = ma.getQualifier()
+    )
+  }
+}
+
+/**
+ * An SSL object that correctly verifies hostnames, or doesn't need to (for instance, because it's a server).
+ *
+ * This is a sanitizer for `SslEndpointIdentificationFlowConfig`.
+ */
+abstract class SslUnsafeCertTrustSanitizer extends DataFlow::Node { }
+
+/**
+ * An `SSLEngine` set in server mode.
+ */
+private class SslEngineServerMode extends SslUnsafeCertTrustSanitizer {
+  SslEngineServerMode() {
+    exists(MethodAccess ma, Method m |
+      m.hasName("setUseClientMode") and
+      m.getDeclaringType().getASupertype*() instanceof SSLEngine and
+      ma.getMethod() = m and
+      ma.getArgument(0).(CompileTimeConstantExpr).getBooleanValue() = false and
+      this.asExpr() = ma.getQualifier()
+    )
+  }
+}
+
+/**
+ * Holds if the return value of `createSocket` is cast to `SSLSocket`
+ * or the qualifier of `createSocket` is an instance of `SSLSocketFactory`.
+ */
+private predicate isSslSocket(MethodAccess createSocket) {
+  exists(Variable ssl, CastExpr ce |
+    ce.getExpr() = createSocket and
+    ce.getControlFlowNode().getASuccessor().(VariableAssign).getDestVar() = ssl and
```

Comment:
Might be better to express the `VariableAssign`-to-`ce` connection using `VariableAssign.getSource() = ce` instead of relying on low-level details of the CFG.

Reply:
Hmm, or just get rid of the `VariableAssign` constraint altogether and do the type test on the `CastExpr`: `ce.getType() instanceof SSLSocket`. That seems simpler and better.

Reply:
Good point, it looks way simpler now. Done in 695e77a, thanks!